book/src/developers/designs/cryptography-key-domains.md

```diff
@@ -0,0 +1,205 @@
+possible.
+
+Entities that rely on a cryptographic key will relate to a Key Object.
+
+```
+  ┌─────────────────────┐
+  │                     │
+  │                     │
+  │    Key Provider     │
+  │                     │
+  │                     │
+  │                     │
+  └─────────────────────┘
+             ▲
+             │
+           ┌─┘
+     ┌─────┼───────────────┐
+    ┌┴─────┼──────────────┐│
+  ┌─┴──────┴────────────┐ ││
+┌─┴───────────────────┐ │ ││
+│                     │ │ ││
+│                     │ │ ││
+│     Key Object      │ │ ││
+│                     │ │ ├┘
+│                     │ ├─┘
+│                     ├─┘
+└─────────────────────┘
+           ▲
+           │
+           │
+           │
+┌─────────────────────┐
+│                     │
+│    Key Consumer     │
+│                     │
+│   * OAuth2 Client   │
+│    * Domain Keys    │
+│                     │
+└─────────────────────┘
+```
+
+Key Objects have a Key Type denoting the type of material they contain. The types will be named
+after the JWA algorithms from [rfc7518](https://www.rfc-editor.org/rfc/rfc7518). This allows easy
+mapping to OAuth2 concepts and PKCS11 in the future.
+
+- `ES256` (ECDSA using P-256 and SHA-256, `CKM_ECDSA_SHA256`)
+- `RS256` (RSASSA-PKCS1-v1\_5 using SHA-256, `CKM_SHA256_RSA_PKCS`)
+- `HS256` (HMAC using SHA-256, `CKM_SHA256_HMAC`)
+
+Possible future classes could be
+
+- `A256GCM` (AES GCM with 256 bit key `CKM_AES_GCM`)
+
+The type defines the possible operations of the Key Object but not how the operation is performed.
+
+A key object MAY have multiple Key Types.
+
+Key Objects also must define their structure related to their Key Provider. For example a possible
+TPM Key Provider needs to store it's Public and Private components in the Key Object, where our
+internal provider needs to store the DER portions of the keys.
+
+Between the type and the provider, this provides a concrete way to determine how a key needs to be
+used.
+
+For each private/public key pair, or each symmetric key, a record of it's status (valid, retained,
+expired, revoked)
+
+Each key may have multiple Key Identifiers. Key Identifiers must be unique.
+
+Every Key Object must have only _one_ key in the valid state. Key rotation causes a key to move to
+the retained state. If a key is missing a valid key, a new one MUST be generated.
+
+On replication revoked, expired, retained and valid take precedence in that order. If two keys are
+marked as valid, the "latest write" wins.
+
+On rotation the private key is _discarded_ to prevent future use of a rotated key.
+
+Keys must be merged, and not deleted.
+
+```
+class: KeyObject
+uuid: ...
+key_object_type: ECDSA_SHA256
+key_object_type: RSA_SHA256
+key_object_type: RSA_SHA256
+es256_private: <private key>
+es256_public: { id: ..., status: valid, public_key }
+hs256_private: <private key>
+rs256_private: <private key>
+rs256_public: { id: ..., status: valid, public_key }
+```
+
```

Comment:
As we discussed, let's add a plan for a "future" key here, it's something that comes up in SAML/other big-system things, where you're going to roll the key and need to tell everyone about it, but not break things yet, so you're going to start on x date (or when the admins agree to enable it) but you need to pre-warn third parties etc. and make sure people are going to trust it straight away.

Another use case would be if you're moving from one key storage to another, you can pre-generate the keys slowly before going live.